packages/service-library/src/servicelib/rabbitmq/rpc_interfaces/async_jobs/async_jobs.py

```diff
@@ -103,3 +121,111 @@ async def submit(
+@retry(**_DEFAULT_RPC_RETRY_POLICY)
+async def _wait_for_completion(
+    rabbitmq_rpc_client: RabbitMQRPCClient,
+    *,
+    rpc_namespace: RPCNamespace,
+    job_id: AsyncJobId,
+    job_id_data: AsyncJobNameData,
+    client_timeout: int,
+) -> AsyncGenerator[AsyncJobStatus, None]:
+    try:
+        async for attempt in AsyncRetrying(
+            stop=stop_after_delay(client_timeout),
+            reraise=True,
+            retry=retry_if_exception_type(TryAgain),
+            before_sleep=before_sleep_log(_logger, logging.DEBUG),
+            wait=wait_fixed(_DEFAULT_POLL_INTERVAL_S),
+        ):
+            with attempt:
+                job_status = await status(
+                    rabbitmq_rpc_client,
+                    rpc_namespace=rpc_namespace,
+                    job_id=job_id,
+                    job_id_data=job_id_data,
+                )
+                yield job_status
+                if not job_status.done:
+                    msg = f"{job_status.job_id=}: '{job_status.progress=}'"
+                    raise TryAgain(msg)  # noqa: TRY301
+
+    except TryAgain as exc:
+        # this is a timeout
+        msg = f"Long running task {job_id=}, calling to  timed-out after {client_timeout} seconds"
+        raise TimeoutError(msg) from exc
+
+
+@dataclass(frozen=True)
+class AsyncJobComposedResult:
+    status: AsyncJobStatus
+    _result: Awaitable[Any] | None = None
+
+    @property
+    def done(self) -> bool:
+        return self._result is not None
+
+    async def result(self) -> Any:
+        if not self._result:
+            msg = "No result ready!"
+            raise ValueError(msg)
+        return await self._result
+
+
+async def submit_and_wait(
+    rabbitmq_rpc_client: RabbitMQRPCClient,
+    *,
+    rpc_namespace: RPCNamespace,
+    method_name: str,
+    job_id_data: AsyncJobNameData,
+    client_timeout: datetime.timedelta,
+    **kwargs,
+) -> AsyncGenerator[AsyncJobComposedResult, None]:
+    async_job_rpc_get = None
+    try:
+        async_job_rpc_get = await submit(
+            rabbitmq_rpc_client,
+            rpc_namespace=rpc_namespace,
+            method_name=method_name,
+            job_id_data=job_id_data,
+            **kwargs,
+        )
+        async for job_status in _wait_for_completion(
+            rabbitmq_rpc_client,
+            rpc_namespace=rpc_namespace,
+            job_id=async_job_rpc_get.job_id,
+            job_id_data=job_id_data,
+            client_timeout=client_timeout,
+        ):
+            yield AsyncJobComposedResult(job_status)
+
+        assert job_status  # nosec
+        yield AsyncJobComposedResult(
+            job_status,
+            result(
+                rabbitmq_rpc_client,
+                rpc_namespace=rpc_namespace,
+                job_id=async_job_rpc_get.job_id,
+                job_id_data=job_id_data,
+            ),
+        )
+    except (TimeoutError, CancelledError):
+        if async_job_rpc_get is not None:
+            await cancel(
```

Comment:
THOUGHT: what if `cancel` raises? 

```python
try:
    try:
        raise OneException("one")
    except OneException:
        raise TwoException("two")  # implicit chaining happens here
except Exception as err:
    import logging
    logging.exception("error happened")
```
shows split in two as `The above exception was the direct cause...`
```
TwoException: two

The above exception was the direct cause of the following exception:

OneException: one
```

but using `from e`  chains them 

```python
except OneException as e:
    raise TwoException("two") from e
```

and will make the log include both tracebacks, with OneException shown as the cause